### PR TITLE
feat: Add API authentication by role and support set default role by Cognito group

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

1. Add API authentication by role
2. Support set default role by Cognito group
3. Delete useless API

## Implementation highlights

**New user login on first time:**

- step1. frontend get user info
- step2. DDB is not exist this user
- step3. get `id_token` from request and decode to json object
- step4. get role from **decode token** use JsonPath that set by environment variable:  `OIDC_ROLE_PATH`

**OIDC_ROLE_PATH** default value is `$.payload.cognito:groups` for Cognito groups.

## Test checklist

- [ ] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/common/constant.ts [hunk 1]
```diff
@@ -226,23 +226,27 @@ export const QUICKSIGHT_USER_NAME_PATTERN =
 export const QUICKSIGHT_NAMESPACE_PATTERN = '^([A-Za-z])[A-Za-z0-9]{4,63}$';
 export const REDSHIFT_DB_NAME_PATTERN = `^${PROJECT_ID_PATTERN}$`;
 export const REDSHIFT_DB_USER_NAME_PATTERN = '^([a-zA-Z][a-zA-Z0-9-_]{1,63})?$';
-export const REDSHIFT_CLUSTER_IDENTIFIER_PATTERN = '^([a-zA-Z][a-zA-Z0-9-_]{1,63})?$';
+export const REDSHIFT_CLUSTER_IDENTIFIER_PATTERN =
```
Comment (zxkane) [inline src/common/constant.ts:229]: Are it and below lines caused by a format setting or an IDE plug-in?

File: src/control-plane/backend/lambda/api/common/utils.ts [hunk 2]
```diff
@@ -106,6 +110,66 @@ function getEmailFromRequestContext(requestContext: string | undefined) {
...
+
+  const OIDC_ROLE_PATH = process.env.OIDC_ROLE_PATH ?? '$.payload.cognito:groups';
+  const values = JSONPath({ path: OIDC_ROLE_PATH, json: decodedToken });
+  if (Array.prototype.isPrototypeOf(values) && values.length > 0) {
+    oidcRoles = values[0] as string[];
+  } else {
+    return role;
+  }
+  if (oidcRoles &&
+          oidcRoles.includes(`Clickstream${IUserRole.OPERATOR}`) &&
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/common/utils.ts:163]: Let's move it configurable via env like `token path`.

File: src/control-plane/backend/lambda/api/common/utils.ts [hunk 3]
```diff
@@ -106,6 +110,66 @@ function getEmailFromRequestContext(requestContext: string | undefined) {
...
+          oidcRoles.includes(`Clickstream${IUserRole.ANALYST}`)) {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/common/utils.ts:164]: Let's move it configurable via env like `token path`.

File: src/control-plane/backend/lambda/api/router/environment.ts [hunk 4]
```diff
@@ -13,20 +13,23 @@
...
+import { defaultAssumeRoleTypeValid, defaultPageValueValid, defaultRegionValueValid, defaultSubnetTypeValid, isRequestIdExisted, isValidEmpty, validate, validateRole } from '../common/request-valid';
+import { IUserRole } from '../common/types';
 import { EnvironmentServ } from '../service/environment';
 
 const router_env = express.Router();
 const environmentServ: EnvironmentServ = new EnvironmentServ();
 
 router_env.get(
   '/regions',
+  validateRole([IUserRole.ADMIN, IUserRole.OPERATOR]),
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/router/environment.ts:25]: Is there a way we do validation in a middleware to check the request path with role without validating in every route?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/lambda/api/router/environment.ts (hunk 4 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/router/environment.ts:25]: if use middleware, also need a `map<api route, roles>` to manage this relationship.

File: src/control-plane/backend/lambda/api/common/utils.ts (hunk 2 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/common/utils.ts:163]: use [this](https://github.com/awslabs/clickstream-analytics-on-aws/pull/205/files#diff-720788bdfb4c591afcdd0e55c0377f7354aaf58151887edc3db3250191c3a246R155) read groups or roles from token

File: src/control-plane/backend/lambda/api/common/utils.ts (hunk 3 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/common/utils.ts:164]: use [this](https://github.com/awslabs/clickstream-analytics-on-aws/pull/205/files#diff-720788bdfb4c591afcdd0e55c0377f7354aaf58151887edc3db3250191c3a246R155) read groups or roles from token

File: src/common/constant.ts (hunk 1 above)
Comment (tyyzqmf) [reply src/common/constant.ts:229]: yarn eslint

File: src/control-plane/backend/lambda/api/common/utils.ts (hunk 2 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/common/utils.ts:163]: Do you mean like `Clickstream${IUserRole.OPERATOR}` set as a env value?

Comment (zxkane) [reply src/control-plane/backend/lambda/api/common/utils.ts:163]: Exactly

File: src/control-plane/backend/lambda/api/router/environment.ts (hunk 4 above)
Comment (zxkane) [reply src/control-plane/backend/lambda/api/router/environment.ts:25]: I think we should also organize the API route with patterns,

- `op/*` for pipeline operation
- `analyze/*` for analyst's explorative analysis
- `users/*` for the administrator to manage the users and roles

Then, we can benefit from it by granting the roles with corresponding permissions.

Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/router/environment.ts]: done

File: src/control-plane/backend/lambda/api/common/utils.ts (hunk 2 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/common/utils.ts]:
```
const OIDC_OPERATOR_ROLE_NAME = process.env.OIDC_OPERATOR_ROLE_NAME ?? `Clickstream${IUserRole.OPERATOR}`;
  const OIDC_ANALYST_ROLE_NAME = process.env.OIDC_ANALYST_ROLE_NAME ?? `Clickstream${IUserRole.ANALYST}`;
```

**Review by zxkane — APPROVED** [review]
Nice!

**Pull request merged**